sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -539,3 +541,82 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
 
   override def prettyName: String = "str_to_map"
 }
+
+/**
+ * Adds/replaces field in struct by name.
+ */
+case class WithFields(
+  structExpr: Expression,
+  nameExprs: Seq[Expression],
+  valExprs: Seq[Expression]) extends Unevaluable {
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val expectedStructType = StructType(Nil).typeName
+    if (structExpr.dataType.typeName != expectedStructType) {
+      TypeCheckResult.TypeCheckFailure(
+        "struct argument should be struct type, got: " + structExpr.dataType.catalogString)
+    } else if (!nameExprs.forall(e => e.foldable && e.dataType == StringType)) {
+      TypeCheckResult.TypeCheckFailure(
+        s"nameExprs argument should contain only foldable ${StringType.catalogString} expressions")
+    } else if (nameExprs.length != valExprs.length) {
```

Comment:
Can we add an assert at the beginning of the class body? This is not a user-facing error that can be triggered by users.

Reply:
Similar to my other comment, I did this because it may be possible for the user to do this when/if we integrate this with the SQL API in the future. I'm happy to change it if you think this isn't something to be concerned about today.

Reply:
yea let's don't consider SQL API for now. But even for SQL API, I think we should follow the scala API and take care of the name parsing, so that users can't trigger this issue.

Reply:
fixed